### ISSUE TITLE
Missing support for array conversion

### ISSUE THREAD

**Issue by hazaart** [NaturalIntelligence/fast-xml-parser#181]
In case a node exists multiple time in the same parent, it will automatically be converted to a JS array.
If the node occurs only once however in the dataset (eventhough it could occur multiple times in theory), the conversion will create an object cause it has no way of knowing it should convert it to an array.
Other libraries have support for this by adding an array in the parsing options containing a list of elements that always need to be converted to an array.
Example https://github.com/abdolence/x2js has support via 'arrayAccessFormPaths'.

I am surprised this basic conceptual issue in xml to json conversion is not yet covered by this widely used library. Makes for a lot of painful checks on the output.

**Comment by amitguptagwl** [comment]
> I am surprised this basic conceptual issue in xml to json conversion is not yet covered by this widely used library. Makes for a lot of painful checks on the output.

:smile: Have you checked for the already raised issues and the discussion on the same topic?

Please give us an example so we can understand if your rasied concern is different.

**Comment by hazaart** [comment]
Hi,

It is the same as issue https://github.com/NaturalIntelligence/fast-xml-parser/issues/166
But it should be supported in the library.
If you allow to pass an optional array which specifies which nodenames should always convert to Array.
See example implementation as mentioned above in x2js.

Being a conceptual issue when making a conversion from XML to JSON, I think it would be nice if a solution is supported by the library.

Thanks

**Comment by amitguptagwl** [comment]
There is no standard for XML to JSON/JS Object conversion. Hence, we need to finalize the response which suits to our implementation and users expectation. People feel its need but no one yet given us the clear requirement.

We don't want hurry. Because many changes cant be reverted once the users at world level start using them.

**Comment by amitguptagwl** [comment]
I'm closing this issue as it is duplicate for the other issue. If you'll feel we should keep it open, please reopen

**Comment by hazaart** [comment]
"People feel its need but no one yet given us the clear requirement."
How do you mean, I just did :-)

In the options we can pass for parsing, add a new optional array. This way there is no conflict with the existing way of working.

arrayAccessFormPaths : [] - Array access paths - use this option to configure paths to XML elements always in "array form". You can configure beforehand paths to all your array elements based on XSD or your knowledge. Every path could be a simple string (like 'parent.child1.child2'), a regex (like /.*.child2/), or a custom function. Default is empty

**Comment by amitguptagwl** [comment]
This is not a clear requirement. I doubt if you have seen the previous discussions.

**Comment by DazChong** [comment]
Yes, this is a missing feature.
Imaging scenario of user object with multiple emails as array, some user have multiple emails and some only have 1.

**Comment by amitguptagwl** [comment]
@hazaart @DazChong FYR https://github.com/NaturalIntelligence/fast-xml-parser/pull/135

**Comment by Kaarel** [comment]
Use cases:
Invoices, stock inventory, accounting, shopping carts, travel reservations, players in a game etc etc. They all have 0 to n items, in practice 1 to n. Basically almost every use case :) 1 is just a special case of n element array.

We have a lot of use cases like below. We do not control the XML that is sent to us, we just need to work with it.

```xml
<report>
    <store>
        <region>US</region>
        <inventory>
            <item>
                <name>Banana</name>
                <count>200</count>
            </item>
            <item>
                <name>Apple</name>
                <count>100</count>
            </item>
        </inventory>
    </store>
    <store>
        <region>EU</region>
        <inventory>
            <item>
                <name>Banana</name>
                <count>100</count>
            </item>
        </inventory>
    </store>
</report>
```

which `fast-xml-parser` parses into

```json
{
  "report": {
    "store": [
      {
        "region": "US",
        "inventory": {
          "item": [
            {
              "name": "Banana",
              "count": "200"
            },
            {
              "name": "Apple",
              "count": "100"
            }
          ]
        }
      },
      {
        "region": "EU",
        "inventory": {
          "item": {
            "name": "Banana",
            "count": "100"
          }
        }
      }
    ]
  }
}
```

Without this feature accessing inventory items for every store requires always testing if `store.inventory.item` is array or not. Same for iterating over stores as well, as there could be only one store sometimes. In practice there are a lot more elements like that in the XML we parse. This results in an unnecessarily complicated application code.

Basically this is the reason we can not use `fast-xml-parser`.

**Comment by amitguptagwl** [comment]
Thanks for highlighting this @Kaarel . So if let's suppose, we have some option which basically renders a nested tag always an array, it'll look like as follow;

```json
{
    "report": [
        {
            "store": [
                {
                    "region": "US",
                    "inventory": [
                        {
                            "item": [
                                {
                                    "name": "Banana",
                                    "count": "200"
                                },
                                {
                                    "name": "Apple",
                                    "count": "100"
                                }
                            ]
                        }
                    ]
                },
                {
                    "region": "EU",
                    "inventory": [
                        {
                            "item": [
                                {
                                    "name": "Banana",
                                    "count": "100"
                                }
                            ]
                        }
                    ]
                }
            ]
        }
    ]
}
```

will it solve the purpose?

**Comment by amitguptagwl** [comment]
I have started the discussion here;

https://github.com/NaturalIntelligence/fast-xml-parser/issues/190

**Comment by Kaarel** [comment]
Yes, specifically the strict mode in the examples of #190

**Comment by gknapp** [comment]
I'm using openlayers, it expects single values be a collection (array) and not unwrapped (object). It checks the parent node length but fails when the single value is unwrapped. Having a configuration option for this would mean I could use fast-xml-parser.

**Comment by amitguptagwl** [comment]
@gknapp Have u looked at #190 ? does it solve ur purpose? if not please give an example of your expectation.

**Comment by gknapp** [comment]
@amitguptagwl Yes that's exactly what's needed. I knew it was coming but didn't think it was available on npm yet?

**Comment by amitguptagwl** [comment]
It is available on Github. I'll publish that on Github tomorrow or on weekend.

**Comment by hazaart** [comment]
I don't think it is wishfull that every single node is transformed into an array. We should be able to control which node names have to always convert to array.

You may have levels in the XML that will only occur once. Take the <report/> tag for example in the above XML.
Let's say you wish to repeat over the stores you need to access it via `report[0].stores`. Having to add the index to the `report `makes for unreadable code and it becomes difficult to see which nodes can in reality recur and which ones don't.

A desired solution would be that the user can pass in an array of nodenames which should always be transformed to array.

Example: 

```
const jsObj = parser.parse(xmlData, {
      arrayNodes: ["store", "item"]
});
```

**Comment by amitguptagwl** [comment]
I'm not sure how many users wish to use `arrayMode` in this way, but this can be added in a future version if more users feel its need.
